can/bit_timing.py

```diff
@@ -83,6 +99,74 @@ def __init__(
         if nof_samples not in (1, 3):
             raise ValueError("nof_samples must be 1 or 3")
 
+        self._sample_point = sample_point
+
+    def _calc_timings(self):
+        if self._f_clock is None:
+            raise ValueError('The f_clock is needed in order to calculate the timings')
+
+        sample = 0
+        res = None
+        last_bitrate = 0
+
+        def _get_calc(error_margin):
+            high_bitrate = int(self._bitrate * (1.0 + (error_margin / 100.0)))
+            low_bitrate = int(self._bitrate * (1.0 - (error_margin / 100.0)))
+            return [
+                (c_brp, c_sjw, c_tseg1, c_tseg2)
+                for c_brp in range(1, 65)  # baudrate prescalar (1-64)
+                for c_sjw in range(1, 5)  # synchronization jump width (1-4)
+                for c_tseg1 in range(1, 17)  # prop_seg + phase_seg1 (1-16)
+                for c_tseg2 in range(1, 9)  # phase_seg2 (1-8)
+                if (
+                    low_bitrate < int(self._f_clock / (c_brp * (1 + c_tseg1 + c_tseg2))) < high_bitrate and
+                    (self._brp is None or c_brp == self._brp) and
+                    (self._sjw is None or c_sjw == self._sjw) and
+                    (self._tseg1 is None or c_tseg1 == self._tseg1) and
+                    (self._tseg2 is None or c_tseg2 == self._tseg2)
+                )
+            ]
+
+        calc = _get_calc(0)
+        if not calc:
+            calc = _get_calc(self.calc_error_margin)
+
+        for can_brp, can_sjw, can_tseg1, can_tseg2 in calc:
+            s = ((can_tseg1 + 1) / (1 + can_tseg1 + can_tseg2)) * 100.0
+
+            if self._sample_point is None:
+                bitrate = int(self._f_clock / (can_brp * (1 + can_tseg1 + can_tseg2)))
+
+                if s < sample:
+                    continue
+                if (
+                    not last_bitrate or
+                    self._bitrate - last_bitrate <= self._bitrate - bitrate
+                ):
+                    last_bitrate = bitrate
+                    sample = s
+                    res = (
+                        can_brp,
+                        can_sjw,
+                        can_tseg1,
+                        can_tseg2
+                    )
+
+            elif s == self._sample_point:
+                res = (
+                    can_brp,
+                    can_sjw,
+                    can_tseg1,
+                    can_tseg2
+                )
+                break
+
+        if res is None:
+            return
+
+        self._brp, self._sjw, self._tseg1, self._tseg2 = res
+        self._bitrate = int(self._f_clock / (self._brp * (1 + self._tseg1 + self._tseg2)))
```

Comment:
don't want to do it that way because the bitrate having a value other then None is what launches the calculations if any of the other timing values are missing.

The only reason why this piece of code is there is because not every bitrate is available, take 33000 as an example. There are no timing values that equal that bitrate . so the closest one to it is i believe 33057 or something along those line. So I coded in a mechanism that allows for a user settable tolerance that gets applied to the initially supplied bitrate so timing values will be used for the closest bitrate. 

I believe that the sjw parameter is what one can use to equalize a non matching bitrate. so the further apart the bus bitrate is from the interface bitrate is the higher the sjw should be set. I believe this is how it works. I haven't located a CAN specification for dummies book LOL. so there is a learning curve with all of the geek speak in the specifications I have come across. There is what appears to be quite a bit of information that bumps heads as well.

So if my understanding is incorrect with respect to the sjw let me know. I believe it also allows for devices that have different OSC frequencies to work on the same bus, this would also cause a change in the bitrate as well, that is why I think that the sjw can be used as a way to handle an interfaces bitrate not matching the buses bitrate exactly. It is also the reason why I i keep the parameters a user has supplied and find matches using those parameters.

Reply:
I have no idea what the CiA recommendations are, so I am not able to answer your questions.

Reply:
I have thought about being able to pass an instance of this class to an interface and that interface would be able to fill in the f_clock and then grab the timings it needs. This way all a user would have to do is construct the BitTiming instance with the bitrate and nothing else.

I do not know what all of the interfaces OSC frequencies are I only know the one for the CANalyst-II which is 16KHz. I also figured that those changes can me made as a pull request for each interface.

The name of the game is provide the most flexability as easily as possible. I have a can network I am connecting my interface to that has a 33000 bitrate and there is no 33000 bitrate available in the `TIMING_DICT` all of the documentation I have come across for the CANalyst-II has the same list of bitrates and timings that is in the `TIMING_DICT`. So I needed a way to get the BTR0 and BTR1 (Timing0 and Timing1) for 33000 bps. That is what this PR stems from. I had already written the core of it when I was dinking about with a clone of the CANalyst-II. that clone had an added function to the ControlCan library where the brp, sjw, tseg1 and tseg2 values could be passed to initialize the channel. The strange thing is that it had a brp that could be in the range of 1 and 1024 (I think that is the high number) where as everything I have seen states 1 - 64. I am not sure why they used 1024.. maybe for CAN FD? IDK...

Reply:
I did a test changing the range for the brp.
It allows for a tone more bitrates. there are a slew of bitrates that have the same btr0 and btr1, the bitrates that share the same btr0 and btr1 are in close proximity to each other.

It also adds abunch of bitrates that are below 5kbps, I didn't use anything below 5kbps.
so here is what the brp ranges map out to with respect to the number of unique bitrates and the number of unique timings (btr0, btr1)

brp range | num bitrates | num timings
----------|------------|--------
1-64 | 668 | 32768
1-1024 | 1211 | 34548
1-500 | 1207 | 34464
1-400 | 1190 | 34464
1-351 | 1174 | 34464
1-300 | 1154 | 33984
1-200 | 1089 | 32768
1-150 | 1012 | 32768

so i could see using the range from 1-351 to offer the largest number of available timings while also having a larger pool of bitrates to match to.

This would allow for the closest possible match. it would have to be tested to see if it would work.

Reply:
The ranges depend on the controller. If you look at https://github.com/linux-can/can-utils/blob/master/can-calc-bit-timing.c you start to realize how complex it is to cover everything. We may need to limit ourselves to SJA1000 compatible controllers (ones that use BTR registers) to make it manageable.

Reply:
let me know what you want done and i will do it.

Reply:
I would say only focus on SJA1000 for now. If sample point has been omitted, default to CiA recommended values. If it is possible to include the rules for SJW that would be great as well.

Reply:
and where would I locate the CiA recommended values and also the rules for sjw. You are speaking Greek to me right now. I just started working with CAN networking 2-3 weeks ago.

Reply:
You can use the code from can-utils above as a reference.

Reply:
OK cool, I have a better understanding of what is happening now. I converted the bit timing calculator from can-utils to Python so I should not have any issue with bringing it into what is already made. It is a a far better way to go about it then using the list comprehension and iteration over all of the values like I am. It is also a bucket load faster.

Reply:
Oh and after this PR is finished up the individual interface classes are going to have to be updated so they will feed in the needed information about the interface so the calculations will be correct. and to also grabe the proper values that are needed to initialize the interface. I could use a hand doing that if you are willing.

Reply:
OK sir, I just made a commit. you just have to run the bit_timings.py file and it will print out an example.